File: mobile/scripts/gen-store-frames.js
#!/usr/bin/env node
/* Builds the framed App Store panels from the raw device captures.
 *
 * The raw set under marketing/ios/6.9/raw is the source of truth: straight
 * simctl framebuffer grabs at 1320x2868, which is exactly the 6.9" slot App
 * Store Connect asks for. This composites each one onto the editorial cream
 * surface with a Fraunces headline, and writes the result at the same size, so
 * either set can be uploaded without resizing anything.
 *
 * The headline copy lives here rather than in a JSON file beside it: it is five
 * lines, and a caption that drifts from the screenshot under it is the failure
 * this file exists to prevent. Read them next to the shots they sit on.
 *
 * Rendering is done by a Python helper (Pillow) because the brand faces ship as
 * variable woff2 and Pillow is the one thing on this machine that will rasterise
 * a variable axis. Node does the orchestration and the copy; Python does pixels.
 */
"use strict";
const { execFileSync } = require("node:child_process");
const fs = require("node:fs");
const path = require("node:path");

const ROOT = path.resolve(__dirname, "..");
const RAW = path.join(ROOT, "marketing/ios/6.9/raw");

/* Two targets, because the stores do not take the same picture.
 *
 * App Store Connect accepts the device capture itself, so the iOS panel keeps
 * 1320x2868 and nothing is resized on the way up.
 *
 * Play caps a phone screenshot at 2:1. The raw capture is 2868/1320 = 2.17:1
 * and would be refused, which is the kind of thing you find out at the end of
 * an upload. Its panel is a 16:9 canvas with the same capture inset — a
 * reframing rather than a crop, so nothing in the shot is lost. */
const TARGETS = [
  {
    label: "App Store 6.9\"",
    out: path.join(ROOT, "marketing/ios/6.9/framed"),
    w: 1320, h: 2868, margin: 96,
    rule_y: 196, rule_w: 76,
    head_size: 92, head_y: 252, head_step: 104,
    sub_size: 36, sub_gap: 24, sub_step: 52,
    shot_w: 1040, shot_top: 700,
  },
  {
    label: "Play phone 16:9",
    out: path.join(ROOT, "marketing/android/phone"),
    w: 1080, h: 1920, margin: 78,
    rule_y: 150, rule_w: 62,
    head_size: 72, head_y: 196, head_step: 82,
    sub_size: 29, sub_gap: 20, sub_step: 42,
    shot_w: 700, shot_top: 470,
  },
];

/* No em dashes, no emojis, and nothing that describes the app as an "AI".
   Each line says what the surface under it actually does. */
const PANELS = [
  { file: "01-chat.png", head: "Your operator,\nin your pocket.", sub: "Ask it to reason, look things up, and keep track of what you are working on." },
  { file: "02-projects.png", head: "Every thread,\nand who answered it.", sub: "Coding, development and research, run by the operator." },
  { file: "03-studio.png", head: "Film, music,\nand the studio.", sub: "The creative house, under one roof." },
  { file: "04-cultivation.png", head: "The grower's\nspace.", sub: "Questions go straight to the mycology expert, with no need to phrase them for a router." },
  { file: "05-panels.png", head: "Operator control,\nfrom the phone.", sub: "See what is running, and stop it from here." },
];

const PY = path.join(__dirname, "gen-store-frames.py");

function main() {
  const missing = PANELS.filter((p) => !fs.existsSync(path.join(RAW, p.file)));
  if (missing.length) {
    console.error(`missing raw captures: ${missing.map((m) => m.file).join(", ")}`);
    console.error(`capture them into ${RAW} first`);
    process.exit(1);
  }
  for (const t of TARGETS) fs.mkdirSync(t.out, { recursive: true });
  const spec = JSON.stringify({ raw: RAW, targets: TARGETS, panels: PANELS });
  const res = execFileSync("python3", [PY], { input: spec, encoding: "utf8" });
  process.stdout.write(res);
}

main();

File: mobile/scripts/gen-store-frames.py
#!/usr/bin/env python3
"""Rasterises the framed store panels. Driven by gen-store-frames.js.

Reads a JSON spec on stdin: {raw, targets:[{out, ...geometry}], panels:[...]}.

Two targets, because the stores do not accept the same picture. App Store
Connect takes the device capture itself at 1320x2868, so the framed iOS panel
keeps that size. Play caps a phone screenshot at 2:1 and the raw capture is
2.17:1, so it would be refused as-is; the Play panel is a shorter canvas with
the same capture inset, which is a reframing rather than a crop.

The brand faces ship as variable woff2, which Pillow cannot open, so they are
converted to TTF in a temp dir and the weight axis is set explicitly. Falling
back to a default instance silently would ship a headline at the wrong weight
and read as a rendering bug rather than a missing step, so a failure to set the
axis is reported.
"""
import json
import os
import sys
import tempfile

from PIL import Image, ImageDraw, ImageFilter, ImageFont
from fontTools.ttLib import TTFont

PAPER = (247, 243, 234)
INK = (26, 23, 20)
DIM = (107, 100, 87)
GOLD = (184, 137, 58)

HERE = os.path.dirname(os.path.abspath(__file__))
FONTS = os.path.normpath(os.path.join(HERE, "..", "..", "assets", "fonts"))


def load_face(woff2_name, ttf_name, weight, size, tmpdir):
    src = os.path.join(FONTS, woff2_name)
    dst = os.path.join(tmpdir, ttf_name)
    if not os.path.exists(dst):
        f = TTFont(src)
        f.flavor = None
        f.save(dst)
    font = ImageFont.truetype(dst, size)
    try:
        font.set_variation_by_axes([weight])
    except Exception as exc:  # noqa: BLE001 - reported, not swallowed
        print(f"  warning: could not set weight {weight} on {ttf_name}: {exc}", file=sys.stderr)
    return font


def wrap(draw, text, font, max_w):
    words, lines, cur = text.split(), [], ""
    for word in words:
        trial = f"{cur} {word}".strip()
        if draw.textlength(trial, font=font) <= max_w:
            cur = trial
        else:
            if cur:
                lines.append(cur)
            cur = word
    if cur:
        lines.append(cur)
    return lines


def rounded_mask(size, radius):
    mask = Image.new("L", size, 0)
    ImageDraw.Draw(mask).rounded_rectangle([0, 0, size[0] - 1, size[1] - 1], radius, fill=255)
    return mask


def build(panel, raw_dir, t, tmpdir):
    W, H = t["w"], t["h"]
    margin = t["margin"]
    canvas = Image.new("RGB", (W, H), PAPER)
    draw = ImageDraw.Draw(canvas)

    head_font = load_face("fraunces-var.woff2", "Fraunces.ttf", 600, t["head_size"], tmpdir)
    sub_font = load_face("inter-var.woff2", "Inter.ttf", 400, t["sub_size"], tmpdir)

    # A short gold rule, the one piece of accent on the panel.
    draw.rectangle([margin, t["rule_y"], margin + t["rule_w"], t["rule_y"] + 4], fill=GOLD)

    y = t["head_y"]
    for line in panel["head"].split("\n"):
        draw.text((margin, y), line, font=head_font, fill=INK)
        y += t["head_step"]

    y += t["sub_gap"]
    for line in wrap(draw, panel["sub"], sub_font, W - margin * 2):
        draw.text((margin, y), line, font=sub_font, fill=DIM)
        y += t["sub_step"]

    shot = Image.open(os.path.join(raw_dir, panel["file"])).convert("RGB")
    sw = t["shot_w"]
    sh = round(sw * shot.height / shot.width)
    shot = shot.resize((sw, sh), Image.LANCZOS)
    # The device corner radius, carried through the same scale the shot was.
    corner = round(186 * sw / 1320)
    mask = rounded_mask((sw, sh), corner)
    x = (W - sw) // 2
    top = t["shot_top"]

    # Tinted shadow, never pure black: a black shadow under warm paper reads as
    # a hole rather than as lift.
    shadow = Image.new("RGBA", (W, H), (0, 0, 0, 0))
    ImageDraw.Draw(shadow).rounded_rectangle(
        [x, top + round(26 * sw / 1040), x + sw, top + sh], corner, fill=(26, 23, 20, 74)
    )
    shadow = shadow.filter(ImageFilter.GaussianBlur(max(12, round(38 * sw / 1040))))
    canvas.paste(Image.alpha_composite(canvas.convert("RGBA"), shadow).convert("RGB"), (0, 0))
    canvas.paste(shot, (x, top), mask)

    dst = os.path.join(t["out"], panel["file"])
    canvas.save(dst)
    return dst, (W, H)


def main():
    spec = json.load(sys.stdin)
    with tempfile.TemporaryDirectory() as tmpdir:
        for t in spec["targets"]:
            os.makedirs(t["out"], exist_ok=True)
            print(f"{t['label']} -> {t['w']}x{t['h']}  (ratio {t['h'] / t['w']:.2f}:1)")
            for panel in spec["panels"]:
                dst, size = build(panel, spec["raw"], t, tmpdir)
                im = Image.open(dst)
                assert im.size == (t["w"], t["h"]), f"{dst} came out {im.size}, not {(t['w'], t['h'])}"
                print(f"  {os.path.basename(dst)}  {im.size[0]}x{im.size[1]}")
    print("done")


if __name__ == "__main__":
    main()
